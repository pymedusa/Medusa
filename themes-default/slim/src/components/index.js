export { default as AddRecommended } from './add-recommended.vue';
export { default as AddShowOptions } from './add-show-options.vue';
export { default as AddShows } from './add-shows.vue';
export { default as AnidbReleaseGroupUi } from './anidb-release-group-ui.vue';
export { default as AppFooter } from './app-footer.vue';
export { default as AppHeader } from './app-header.vue';
export { default as Backstretch } from './backstretch.vue';
export { default as Config } from './config.vue';
export { default as ConfigGeneral } from './config-general.vue';
export { default as ConfigPostProcessing } from './config-post-processing.vue';
export { default as ConfigNotifications } from './config-notifications.vue';
export { default as ConfigSearch } from './config-search.vue';
export { default as DisplayShow } from './display-show.vue';
export { default as EditShow } from './edit-show.vue';
export { default as History } from './history.vue';
export { default as Home } from './home.vue';
export { default as IRC } from './irc.vue';
export { default as Login } from './login.vue';
export { default as Logs } from './logs.vue';
export { default as ManualPostProcess } from './manual-post-process.vue';
export { default as RootDirs } from './root-dirs.vue';
export { default as Schedule } from './schedule.vue';
export { default as ShowHeader } from './show-header.vue';
export { default as ShowHistory } from './show-history.vue';
export { default as ShowResults } from './show-results.vue';
export { default as SnatchSelection } from './snatch-selection.vue';
export { default as Status } from './status.vue';
export { default as SubMenu } from './sub-menu.vue';
export { default as SubtitleSearch } from './subtitle-search.vue';
export * from './http';
export * from './helpers';
