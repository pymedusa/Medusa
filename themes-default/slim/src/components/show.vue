<script>
import { mapState, mapGetters, mapActions } from 'vuex';
import { apiRoute } from '../api';
import { AppLink, PlotInfo } from './helpers';
import ShowHeader from './show-header.vue';

export default {
    name: 'show',
    template: '#show-template',
    components: {
        AppLink,
        PlotInfo,
        ShowHeader
    },
    metaInfo() {
        if (!this.show || !this.show.title) {
            return {
                title: 'Medusa'
            };
        }
        const { title } = this.show;
        return {
            title,
            titleTemplate: '%s | Medusa'
        };
    },
    props: {
        /**
         * Show indexer
         */
        showIndexer: {
            type: String
        },
        /**
         * Show id
         */
        showId: {
            type: Number
        }
    },
    data() {
        return {};
    },
    computed: {
        ...mapState({
            shows: state => state.shows.shows
        }),
        ...mapGetters({
            getShowById: 'getShowById',
            show: 'getCurrentShow'
        }),
        indexer() {
            return this.showIndexer || this.$route.query.indexername;
        },
        id() {
            return this.showId || Number(this.$route.query.seriesid) || undefined;
        }
    },
    mounted() {
        const {
            id,
            indexer,
            getShow,
            setEpisodeSceneNumbering,
            setAbsoluteSceneNumbering,
            setInputValidInvalid,
            getSeasonSceneExceptions,
            $store,
            show
        } = this;

        // Let's tell the store which show we currently want as current.
        $store.commit('currentShow', {
            indexer,
            id
        });

        // We need the episodes, so let's get it.
        // Enable detailed as we need the xemNumbering for the show-header.
        if (!show || !show.seasons) {
            getShow({ id, indexer, detailed: true, episodes: true });
        }

        this.$watch('show', () => {
            this.$nextTick(() => this.reflowLayout());
        });

        ['load', 'resize'].map(event => {
            return window.addEventListener(event, () => {
                this.reflowLayout();
            });
        });

        window.addEventListener('load', () => {
            $.ajaxEpSearch({
                colorRow: true
            });

            startAjaxEpisodeSubtitles(); // eslint-disable-line no-undef
            $.ajaxEpSubtitlesSearch();
            $.ajaxEpRedownloadSubtitle();
        });

        $(document.body).on('click', '.seasonCheck', event => {
            const seasCheck = event.currentTarget;
            const seasNo = $(seasCheck).attr('id');

            $('#collapseSeason-' + seasNo).collapse('show');
            const seasonIdentifier = 's' + seasNo;
            $('.epCheck:visible').each((index, element) => {
                const epParts = $(element).attr('id').split('e');
                if (epParts[0] === seasonIdentifier) {
                    element.checked = seasCheck.checked;
                }
            });
        });

        let lastCheck = null;
        $(document.body).on('click', '.epCheck', event => {
            const target = event.currentTarget;
            if (!lastCheck || !event.shiftKey) {
                lastCheck = target;
                return;
            }

            const check = target;
            let found = 0;

            $('.epCheck').each((index, element) => {
                if (found === 1) {
                    element.checked = lastCheck.checked;
                }

                if (found === 2) {
                    return false;
                }

                if (element === check || element === lastCheck) {
                    found++;
                }
            });
        });

        // Initially show/hide all the rows according to the checkboxes
        $('#checkboxControls input').each((index, element) => {
            const status = $(element).prop('checked');
            $('tr.' + $(element).attr('id')).each((index, tableRow) => {
                if (status) {
                    $(tableRow).show();
                } else {
                    $(tableRow).hide();
                }
            });
        });

        $(document.body).on('change', '.sceneSeasonXEpisode', event => {
            const target = event.currentTarget;
            // Strip non-numeric characters
            const value = $(target).val();
            $(target).val(value.replace(/[^0-9xX]*/g, ''));
            const forSeason = $(target).attr('data-for-season');
            const forEpisode = $(target).attr('data-for-episode');

            // If empty reset the field
            if (value === '') {
                setEpisodeSceneNumbering(forSeason, forEpisode, null, null);
                return;
            }

            const m = $(target).val().match(/^(\d+)x(\d+)$/i);
            const onlyEpisode = $(target).val().match(/^(\d+)$/i);
            let sceneSeason = null;
            let sceneEpisode = null;
            let isValid = false;
            if (m) {
                sceneSeason = m[1];
                sceneEpisode = m[2];
                isValid = setInputValidInvalid(true, $(target));
            } else if (onlyEpisode) {
                // For example when '5' is filled in instead of '1x5', asume it's the first season
                sceneSeason = forSeason;
                sceneEpisode = onlyEpisode[1];
                isValid = setInputValidInvalid(true, $(target));
            } else {
                isValid = setInputValidInvalid(false, $(target));
            }

            if (isValid) {
                setEpisodeSceneNumbering(forSeason, forEpisode, sceneSeason, sceneEpisode);
            }
        });

        $(document.body).on('change', '.sceneAbsolute', event => {
            const target = event.currentTarget;
            // Strip non-numeric characters
            $(target).val($(target).val().replace(/[^0-9xX]*/g, ''));
            const forAbsolute = $(target).attr('data-for-absolute');

            const m = $(target).val().match(/^(\d{1,3})$/i);
            let sceneAbsolute = null;
            if (m) {
                sceneAbsolute = m[1];
            }
            setAbsoluteSceneNumbering(forAbsolute, sceneAbsolute);
        });

        $('#showTable, #animeTable').tablesorter({
            widgets: ['saveSort', 'stickyHeaders', 'columnSelector'],
            widgetOptions: {
                columnSelector_saveColumns: true, // eslint-disable-line camelcase
                columnSelector_layout: '<label><input type="checkbox">{name}</label>', // eslint-disable-line camelcase
                columnSelector_mediaquery: false, // eslint-disable-line camelcase
                columnSelector_cssChecked: 'checked' // eslint-disable-line camelcase
            }
        });

        $('#popover').popover({
            placement: 'bottom',
            html: true, // Required if content has HTML
            content: '<div id="popover-target"></div>'
        }).on('shown.bs.popover', () => { // Bootstrap popover event triggered when the popover opens
            $.tablesorter.columnSelector.attachTo($('#showTable, #animeTable'), '#popover-target');
        });

        // Changes the button when clicked for collapsing/expanding the season to show/hide episodes
        document.querySelectorAll('.collapse.toggle').forEach(element => {
            element.addEventListener('hide.bs.collapse', () => {
                // On hide
                const reg = /collapseSeason-(\d+)/g;
                const result = reg.exec(this.id);
                $('#showseason-' + result[1]).text('Show Episodes');
                $('#season-' + result[1] + '-cols').addClass('shadow');
            });
            element.addEventListener('show.bs.collapse', () => {
                // On show
                const reg = /collapseSeason-(\d+)/g;
                const result = reg.exec(this.id);
                $('#showseason-' + result[1]).text('Hide Episodes');
                $('#season-' + result[1] + '-cols').removeClass('shadow');
            });
        });

        // Get the season exceptions and the xem season mappings.
        getSeasonSceneExceptions();
    },
    methods: {
        ...mapActions({
            getShow: 'getShow' // Map `this.getShow()` to `this.$store.dispatch('getShow')`
        }),
        /**
         * Attaches IMDB tooltip,
         * Moves summary and checkbox controls backgrounds
         */
        reflowLayout() {
            console.debug('Reflowing layout');

            this.$nextTick(() => {
                this.moveSummaryBackground();
                this.movecheckboxControlsBackground();
            });

            attachImdbTooltip(); // eslint-disable-line no-undef
        },
        /**
         * Adjust the summary background position and size on page load and resize
         */
        moveSummaryBackground() {
            const height = $('#summary').height() + 10;
            const top = $('#summary').offset().top + 5;

            $('#summaryBackground').height(height);
            $('#summaryBackground').offset({ top, left: 0 });
            $('#summaryBackground').show();
        },
        /**
         * Adjust the checkbox controls (episode filter) background position
         */
        movecheckboxControlsBackground() {
            const height = $('#checkboxControls').height() + 10;
            const top = $('#checkboxControls').offset().top - 3;

            $('#checkboxControlsBackground').height(height);
            $('#checkboxControlsBackground').offset({ top, left: 0 });
            $('#checkboxControlsBackground').show();
        },
        setEpisodeSceneNumbering(forSeason, forEpisode, sceneSeason, sceneEpisode) {
            const indexerName = $('#indexer-name').val();
            const seriesId = $('#series-id').val();

            if (sceneSeason === '') {
                sceneSeason = null;
            }
            if (sceneEpisode === '') {
                sceneEpisode = null;
            }

            $.getJSON('home/setSceneNumbering', {
                indexername: indexerName,
                seriesid: seriesId,
                forSeason,
                forEpisode,
                sceneSeason,
                sceneEpisode
            }, data => {
                // Set the values we get back
                if (data.sceneSeason === null || data.sceneEpisode === null) {
                    $('#sceneSeasonXEpisode_' + seriesId + '_' + forSeason + '_' + forEpisode).val('');
                } else {
                    $('#sceneSeasonXEpisode_' + seriesId + '_' + forSeason + '_' + forEpisode).val(data.sceneSeason + 'x' + data.sceneEpisode);
                }
                if (!data.success) {
                    if (data.errorMessage) {
                        alert(data.errorMessage); // eslint-disable-line no-alert
                    } else {
                        alert('Update failed.'); // eslint-disable-line no-alert
                    }
                }
            });
        },
        setAbsoluteSceneNumbering(forAbsolute, sceneAbsolute) {
            const indexerName = $('#indexer-name').val();
            const seriesId = $('#series-id').val();

            if (sceneAbsolute === '') {
                sceneAbsolute = null;
            }

            $.getJSON('home/setSceneNumbering', {
                indexername: indexerName,
                seriesid: seriesId,
                forAbsolute,
                sceneAbsolute
            }, data => {
                // Set the values we get back
                if (data.sceneAbsolute === null) {
                    $('#sceneAbsolute_' + seriesId + '_' + forAbsolute).val('');
                } else {
                    $('#sceneAbsolute_' + seriesId + '_' + forAbsolute).val(data.sceneAbsolute);
                }

                if (!data.success) {
                    if (data.errorMessage) {
                        alert(data.errorMessage); // eslint-disable-line no-alert
                    } else {
                        alert('Update failed.'); // eslint-disable-line no-alert
                    }
                }
            });
        },
        setInputValidInvalid(valid, el) {
            if (valid) {
                $(el).css({
                    'background-color': '#90EE90', // Green
                    'color': '#FFF', // eslint-disable-line quote-props
                    'font-weight': 'bold'
                });
                return true;
            }
            $(el).css({
                'background-color': '#FF0000', // Red
                'color': '#FFF!important', // eslint-disable-line quote-props
                'font-weight': 'bold'
            });
            return false;
        },
        // @TODO: OMG: This is just a basic json, in future it should be based on the CRUD route.
        // Get the season exceptions and the xem season mappings.
        getSeasonSceneExceptions() {
            const { indexer, id } = this;

            if (!indexer || !id) {
                console.warn('Unable to get season scene exceptions: Unknown series identifier');
                return;
            }

            apiRoute.get('home/getSeasonSceneExceptions', {
                params: {
                    indexername: indexer,
                    seriesid: id
                }
            }).then(response => {
                this.setSeasonSceneExceptions(response.data);
            }).catch(error => {
                console.error('Error getting season scene exceptions', error);
            });
        },
        // Set the season exception based on using the get_xem_numbering_for_show() for animes if available in data.xemNumbering,
        // or else try to map using just the data.season_exceptions.
        setSeasonSceneExceptions(data) {
            $.each(data.seasonExceptions, (season, nameExceptions) => {
                let foundInXem = false;
                // Check if it is a season name exception, we don't handle the show name exceptions here
                if (season >= 0) {
                    // Loop through the xem mapping, and check if there is a xem_season, that needs to show the season name exception
                    $.each(data.xemNumbering, (indexerSeason, xemSeason) => {
                        if (xemSeason === parseInt(season, 10)) {
                            foundInXem = true;
                            $('<img>', {
                                id: 'xem-exception-season-' + xemSeason,
                                alt: '[xem]',
                                height: '16',
                                width: '16',
                                src: 'images/xem.png',
                                title: nameExceptions.join(', ')
                            }).appendTo('[data-season=' + indexerSeason + ']');
                        }
                    });

                    // This is not a xem season exception, let's set the exceptions as a medusa exception
                    if (!foundInXem) {
                        $('<img>', {
                            id: 'xem-exception-season-' + season,
                            alt: '[medusa]',
                            height: '16',
                            width: '16',
                            src: 'images/ico/favicon-16.png',
                            title: nameExceptions.join(', ')
                        }).appendTo('[data-season=' + season + ']');
                    }
                }
            });
        }
    }
};
</script>

<style>

</style>
