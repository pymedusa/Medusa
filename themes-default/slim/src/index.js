import $ from 'jquery';
import 'bootstrap'; // eslint-disable-line import/no-unassigned-import
import 'bootstrap/dist/css/bootstrap.min.css'; // eslint-disable-line import/no-unassigned-import
import './css/open-sans.css'; // eslint-disable-line import/no-unassigned-import

import Vue from 'vue';
import Vuex from 'vuex';
import VueMeta from 'vue-meta';
import VueRouter from 'vue-router';
import VueNativeSock from 'vue-native-websocket';
import AsyncComputed from 'vue-async-computed';
import ToggleButton from 'vue-js-toggle-button';
import Snotify from 'vue-snotify';
import axios from 'axios';
import debounce from 'lodash/debounce';
import store from './store';
import router from './router';
import { isDevelopment } from './utils';
import { apiRoute, apiv1, api, webRoot, apiKey } from './api';
import { Asset, Backstretch, DisplayShow, PlotInfo, ShowSelector, ScrollButtons, FileBrowser, NamePattern, AppHeader, AppLink, SelectList, LanguageSelect, RootDirs, AnidbReleaseGroupUi, SnatchSelection } from './components';

if (window) {
    window.isDevelopment = isDevelopment;

    // Adding libs to window so mako files can use them
    window.$ = $;
    window.jQuery = $;
    window.Vue = Vue;
    window.Vuex = Vuex;
    window.VueMeta = VueMeta;
    window.VueRouter = VueRouter;
    window.VueNativeSock = VueNativeSock;
    window.AsyncComputed = AsyncComputed;
    window.ToggleButton = ToggleButton;
    window.Snotify = Snotify;
    window.axios = axios;
    window._ = { debounce };
    window.store = store;
    window.router = router;
    window.apiRoute = apiRoute;
    window.apiv1 = apiv1;
    window.api = api;

    window.MEDUSA = {
        common: {},
        config: {},
        home: {},
        manage: {},
        history: {},
        errorlogs: {},
        schedule: {},
        addShows: {}
    };
    window.webRoot = webRoot;
    window.apiKey = apiKey;
    window.apiRoot = webRoot + '/api/v2/';

    // Push pages that load via a vue file but still use `el` for mounting
    window.components = [];
    window.components.push(AnidbReleaseGroupUi);
    window.components.push(AppHeader);
    window.components.push(AppLink);
    window.components.push(Asset);
    window.components.push(Backstretch);
    window.components.push(DisplayShow);
    window.components.push(FileBrowser);
    window.components.push(LanguageSelect);
    window.components.push(NamePattern);
    window.components.push(PlotInfo);
    window.components.push(RootDirs);
    window.components.push(ScrollButtons);
    window.components.push(SelectList);
    window.components.push(ShowSelector);
    window.components.push(SnatchSelection);
}
const UTIL = {
    exec(controller, action) {
        const ns = MEDUSA;
        action = (action === undefined) ? 'init' : action;

        if (controller !== '' && ns[controller] && typeof ns[controller][action] === 'function') {
            ns[controller][action]();
        }
    },
    init() {
        if (typeof startVue === 'function') { // eslint-disable-line no-undef
            startVue(); // eslint-disable-line no-undef
        } else {
            $('[v-cloak]').removeAttr('v-cloak');
        }

        const { body } = document;
        $('[asset]').each(function() {
            const asset = $(this).attr('asset');
            const series = $(this).attr('series');
            const path = apiRoot + 'series/' + series + '/asset/' + asset + '?api_key=' + apiKey;
            if (this.tagName.toLowerCase() === 'img') {
                const defaultPath = $(this).attr('src');
                if ($(this).attr('lazy') === 'on') {
                    $(this).attr('data-original', path);
                } else {
                    $(this).attr('src', path);
                }
                $(this).attr('onerror', 'this.src = "' + defaultPath + '"; return false;');
            }
            if (this.tagName.toLowerCase() === 'a') {
                $(this).attr('href', path);
            }
        });
        const controller = body.getAttribute('data-controller');
        const action = body.getAttribute('data-action');

        UTIL.exec('common'); // Load common
        UTIL.exec(controller); // Load MEDUSA[controller]
        UTIL.exec(controller, action); // Load MEDUSA[controller][action]

        window.dispatchEvent(new Event('medusa-loaded'));
    }
};

$.fn.extend({
    addRemoveWarningClass(_) {
        if (_) {
            return $(this).removeClass('warning');
        }
        return $(this).addClass('warning');
    }
});

const { pathname } = window.location;
if (!pathname.includes('/login') && !pathname.includes('/apibuilder')) {
    api.get('config/main').then(response => {
        $.extend(MEDUSA.config, response.data);
        MEDUSA.config.themeSpinner = MEDUSA.config.themeName === 'dark' ? '-dark' : '';
        MEDUSA.config.loading = '<img src="images/loading16' + MEDUSA.config.themeSpinner + '.gif" height="16" width="16" />';

        if (navigator.userAgent.indexOf('PhantomJS') === -1) {
            $(document).ready(UTIL.init);
        }

        MEDUSA.config.indexers.indexerIdToName = indexerId => {
            if (!indexerId) {
                return '';
            }
            return Object.keys(MEDUSA.config.indexers.config.indexers).filter(indexer => { // eslint-disable-line array-callback-return
                if (MEDUSA.config.indexers.config.indexers[indexer].id === parseInt(indexerId, 10)) {
                    return MEDUSA.config.indexers.config.indexers[indexer].name;
                }
            })[0];
        };

        MEDUSA.config.indexers.nameToIndexerId = name => {
            if (!name) {
                return '';
            }
            return MEDUSA.config.indexers.config.indexers[name];
        };
    }).catch(error => {
        console.debug(error);
        alert('Unable to connect to Medusa!'); // eslint-disable-line no-alert
    });
}
