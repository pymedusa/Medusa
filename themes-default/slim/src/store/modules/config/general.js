import { ADD_CONFIG } from '../../mutation-types';
import { arrayUnique, arrayExclude } from '../../../utils/core';

const state = {
    addTitleWithYear: null,
    wikiUrl: null,
    donationsUrl: null,
    namingForceFolders: null,
    sourceUrl: null,
    rootDirs: [],
    brokenProviders: [],
    logs: {
        debug: null,
        dbDebug: null,
        loggingLevels: {},
        numErrors: null,
        numWarnings: null,
        actualLogDir: null,
        nr: null,
        size: null,
        subliminalLog: null,
        privacyLevel: null,
        custom: {}
    },
    cpuPreset: null,
    anonRedirect: null,
    recentShows: [],
    randomShowSlug: null, // @TODO: Recreate this in Vue when the webapp has a reliable list of shows to choose from.
    showDefaults: {
        status: null,
        statusAfter: null,
        quality: null,
        subtitles: null,
        seasonFolders: null,
        anime: null,
        scene: null,
        showLists: null
    },
    launchBrowser: null,
    defaultPage: null,
    trashRemoveShow: null,
    indexerDefaultLanguage: null,
    showUpdateHour: null,
    indexerTimeout: null,
    indexerDefault: null,
    plexFallBack: {
        enable: null,
        notifications: null,
        timeout: null
    },
    recommended: {
        cache: {
            shows: null,
            trakt: null,
            imdb: null,
            anidb: null,
            anilist: null,
            purgeAfterDays: null
        },
        trakt: {
            selectedLists: [],
            availableLists: []
        }
    },
    versionNotify: null,
    autoUpdate: null,
    updateFrequency: null,
    notifyOnUpdate: null,
    availableThemes: null,
    timePresets: [],
    datePresets: [],
    webInterface: {
        apiKey: null,
        log: null,
        username: null,
        password: null,
        port: null,
        host: null,
        notifyOnLogin: null,
        ipv6: null,
        httpsEnable: null,
        httpsCert: null,
        httpsKey: null,
        handleReverseProxy: null
    },
    sslVerify: null,
    sslCaBundle: null,
    noRestart: null,
    encryptionVersion: null,
    calendarUnprotected: null,
    calendarIcons: null,
    proxySetting: null,
    proxyProviders: null,
    proxyClients: null,
    proxyIndexers: null,
    proxyOthers: null,
    skipRemovedFiles: null,
    epDefaultDeletedStatus: null,
    developer: null,
    experimental: null,
    git: {
        username: null,
        password: null,
        token: null,
        authType: null,
        remote: null,
        path: null,
        org: null,
        reset: null,
        resetBranches: null,
        url: null
    },
    // Remove backlogOverview after manage_backlogOverview.mako is gone.
    backlogOverview: {
        status: null,
        period: null
    },
    // Remove themeName when we get fully rid of MEDUSA.config.
    themeName: null,
    providers: {
        prowlarr: {
            url: null,
            apikey: null
        }
    },
    backup: {
        cacheDb: null,
        cacheFiles: null
    }
};

const mutations = {
    [ADD_CONFIG](state, { section, config }) {
        if (section === 'main') {
            state = Object.assign(state, config);
        }
    },
    addRecentShow(state, { show }) {
        state.recentShows = state.recentShows.filter(
            filterShow =>
                !(filterShow.showSlug === show.showSlug && filterShow.name === show.name)
        );

        state.recentShows.unshift(show); // Add the new show object to the start of the array.
        state.recentShows = state.recentShows.slice(0, 5); // Cut the array of at 5 items.
    },
    updateTraktSelectedLists(state, selectedLists) {
        state.recommended.trakt.selectedLists = selectedLists;
    }
};

const getters = {
    effectiveIgnored: (state, _, rootState) => series => {
        const seriesIgnored = series.config.release.ignoredWords.map(x => x.toLowerCase());
        const globalIgnored = rootState.config.search.filters.ignored.map(x => x.toLowerCase());
        if (!series.config.release.ignoredWordsExclude) {
            return arrayUnique(globalIgnored.concat(seriesIgnored));
        }
        return arrayExclude(globalIgnored, seriesIgnored);
    },
    effectiveRequired: (state, _, rootState) => series => {
        const seriesRequired = series.config.release.requiredWords.map(x => x.toLowerCase());
        const globalRequired = rootState.config.search.filters.required.map(x => x.toLowerCase());
        if (!series.config.release.requiredWordsExclude) {
            return arrayUnique(globalRequired.concat(seriesRequired));
        }
        return arrayExclude(globalRequired, seriesRequired);
    }
};

const actions = {
    getConfig({ rootState, commit }, section) {
        return rootState.auth.client.api.get('/config/' + (section || '')).then(res => {
            if (section) {
                const config = res.data;
                commit(ADD_CONFIG, { section, config });
                return config;
            }

            const sections = res.data;
            Object.keys(sections).forEach(section => {
                const config = sections[section];
                commit(ADD_CONFIG, { section, config });
            });
            return sections;
        });
    },
    setConfig({ rootState }, { section, config }) {
        return rootState.auth.client.api.patch(`config/${section}`, config);
    },
    updateConfig({ commit }, { section, config }) {
        return commit(ADD_CONFIG, { section, config });
    },
    getApiKey({ rootState, commit }) {
        const section = 'main';
        const config = { webInterface: { apiKey: '' } };
        return rootState.auth.client.apiRoute.get('config/general/generate_api_key')
            .then(response => {
                config.webInterface.apiKey = response.data;
                return commit(ADD_CONFIG, { section, config });
            });
    },
    setRecentShow({ rootState, commit, state }, show) {
        commit('addRecentShow', { show });
        const config = {
            recentShows: state.recentShows
        };
        return rootState.auth.client.api.patch('config/main', config);
    },
    setCustomLogs({ rootState, commit }, logs) {
        // Convert back to object.
        const reducedLogs = logs.reduce((obj, item) => ({ ...obj, [item.identifier]: item.level }), {});

        return rootState.auth.client.api.patch('config/main', { logs: { custom: logs } })
            .then(() => {
                return commit(ADD_CONFIG, {
                    section: 'main', config: { logs: { custom: reducedLogs } }
                });
            });
    },
    setTraktSelectedLists({ rootState, commit }, selectedLists) {
        return rootState.auth.client.api.patch('config/main', { recommended: { trakt: { selectedLists } } })
            .then(() => {
                return commit('updateTraktSelectedLists', selectedLists);
            });
    }
};

export default {
    state,
    mutations,
    getters,
    actions
};
