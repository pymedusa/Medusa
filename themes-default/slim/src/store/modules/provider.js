import Vue from 'vue';

import { api } from '../../api';
import { ADD_PROVIDERS, ADD_PROVIDER_CACHE, ADD_SEARCH_RESULTS, SET_PROVIDER_CACHE } from '../mutation-types';

const state = {
    providers: {}
};

const mutations = {
    [ADD_PROVIDERS](state, providers) {
        for (const provider of providers) {
            Vue.set(state.providers, provider.id, { ...state.providers[provider.id], ...provider });
        }
    },
    [ADD_PROVIDER_CACHE](state, { providerId, cache }) {
        // Check if this provider has already been added.
        if (!state.providers[providerId]) {
            state.providers[providerId] = {
                name: '',
                config: {},
                cache: []
            };
        }

        Vue.set(state.providers[providerId], 'cache', [...state.providers[providerId].cache || [], ...cache]);
    },
    [SET_PROVIDER_CACHE](state, { providerId, cache }) {
        Vue.set(state.providers[providerId], 'cache', cache);
    },
    /**
     * Add search results which have been retreived through the webSocket.
     * @param {*} state - Vue state
     * @param {Array} searchResults - One or more search results.
     */
    [ADD_SEARCH_RESULTS](state, searchResults) {
        for (const searchResult of searchResults) {
            if (!state.providers[searchResult.provider.id]) {
                state.providers[searchResult.provider.id] = {
                    name: '',
                    config: {},
                    cache: []
                };
            }

            const { cache } = state.providers[searchResult.provider.id];

            // Check if we don't allready have this result in our store.
            // In that case, we update the existing object.
            const existingSearchResult = (cache || []).find(result => result.identifier === searchResult.identifier);
            if (existingSearchResult) {
                // Because this is an existing result, whe're not overwriting dateAdded field.
                const { dateAdded, ...rest } = searchResult;
                Vue.set(state.providers[searchResult.provider.id].cache, cache.indexOf(existingSearchResult), { ...existingSearchResult, ...rest });
            } else {
                Vue.set(state.providers[searchResult.provider.id], 'cache', [...cache || [], ...[searchResult]]);
            }
        }
    }
};

const getters = {};

/**
 * An object representing request parameters for getting a show from the API.
 *
 * @typedef {object} ShowGetParameters
 * @property {boolean} detailed Fetch detailed information? (e.g. scene/xem numbering)
 * @property {boolean} episodes Fetch seasons & episodes?
 */

const actions = {
    /**
     * Get providers.
     *
     * @param {*} context The store context.
     * @returns {Promise} The API response.
     */
    getProviders(context) {
        return new Promise((resolve, reject) => {
            const { commit } = context;
            api.get('/providers')
                .then(response => {
                    commit(ADD_PROVIDERS, response.data);
                    resolve();
                })
                .catch(error => {
                    console.error(`Could not get providers with error: ${error}`);
                    reject();
                });
        });
    },
    /**
     * Get provider cache results for enabled providers.
     *
     * @param {*} context The store context.
     * @param {String} The provider id.
     * @returns {void}.
     */
    async getProviderCacheResults(context, { showSlug, season, episode }) {
        const { commit, state } = context;
        const limit = 1000;
        const params = { limit, showslug: showSlug, season };
        if (episode) {
            params.episode = episode;
        }

        const getProviderResults = async provider => {
            let page = 0;
            let lastPage = false;
            const results = [];

            // Empty the providers cache
            commit(SET_PROVIDER_CACHE, { providerId: provider, cache: [] });

            const { id: providerId } = state.providers[provider];

            page = 0;
            lastPage = false;

            while (!lastPage) {
                let response = null;
                page += 1;

                params.page = page;

                try {
                    response = await api.get(`/providers/${providerId}/results`, { params }); // eslint-disable-line no-await-in-loop
                } catch (error) {
                    if (error.response && error.response.status === 404) {
                        console.debug(`No results available for provider ${provider}`);
                    }

                    lastPage = true;
                }

                if (response) {
                    commit(ADD_PROVIDER_CACHE, { providerId, cache: response.data });
                    results.push(...response.data);

                    if (response.data.length < limit) {
                        lastPage = true;
                    }
                } else {
                    lastPage = true;
                }
            }
            return results;
        };

        const result = {
            providersSearched: 0,
            totalSearchResults: []
        };

        for (const provider in state.providers) {
            if (!state.providers[provider].config.enabled) {
                continue;
            }

            result.providersSearched += 1;
            const providerResults = await getProviderResults(provider); // eslint-disable-line no-await-in-loop
            result.totalSearchResults.push(...providerResults);
        }

        return result;
    },
    /**
     * Get provider cache results for enabled providers.
     *
     * @param {*} {commit} Destructured commit object.
     * @param {Object} searchResult - Search result.
     * @returns {void}.
     */
    addManualSearchResult({ commit }, searchResult) {
        commit(ADD_SEARCH_RESULTS, [searchResult]);
    }

};

export default {
    state,
    mutations,
    getters,
    actions
};
