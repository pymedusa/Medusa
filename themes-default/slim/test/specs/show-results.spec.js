import Vuex, { Store } from 'vuex';
import VueRouter from 'vue-router';
import vueCookies from 'vue-cookies';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import { ShowResults } from '../../src/components';
import general from '../../src/store/modules/config/general';
import historyModule from '../../src/store/modules/history';
import providerModule from '../../src/store/modules/provider';
import show from '../__fixtures__/show-detailed';
import episodeHistory from '../__fixtures__/episode-history';
import fixtures from '../__fixtures__/common';
import provider from '../__fixtures__/providers';

describe('ShowResults.test.js', () => {
    let localVue;
    let store;
    let { state } = fixtures;
    state = {
        ...state,
        ...{ provider },
        ...{ history: {
            history: [],
            page: 0,
            showHistory: {},
            episodeHistory: {
                tvdb5692427: {
                    s04e06: episodeHistory
                }
            }
        } },
        ...{ search: {
            queueitems: []
        } }
    };

    beforeEach(() => {
        localVue = createLocalVue();
        localVue.use(Vuex);
        localVue.use(VueRouter);
        localVue.use(vueCookies);

        store = new Store({
            modules: {
                general: {
                    getters: general.getters,
                    state: state.config.general
                },
                history: {
                    getters: historyModule.getters,
                    state: state.history
                },
                provider: {
                    getters: providerModule.getters,
                    state: state.provider,
                    actions: {
                        getProviders: jest.fn(),
                        getProviderCacheResults: jest.fn().mockImplementation(() => Promise.resolve({
                            providersSearched: 0,
                            totalSearchResults: []
                        }))
                    }
                },
                config: {
                    state: state.config
                },
                search: {
                    state: state.search
                }
            }
        });
    });

    it('renders show-results component', async () => {
        const wrapper = shallowMount(ShowResults, {
            localVue,
            store,
            propsData: {
                show,
                season: 4,
                episode: 6
            }
        });

        await wrapper.vm.$nextTick();
        expect(wrapper.element).toMatchSnapshot();
    });
});
