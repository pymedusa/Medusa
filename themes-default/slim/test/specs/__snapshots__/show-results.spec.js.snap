// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowResults.test.js renders show-results component 1`] = `
<div
  class="show-results-wrapper"
>
  <div
    class="row horizontal-scroll fanartBackground"
  >
    <div
      class="col-md-12 top-15"
    >
      <div
        class="button-row"
      >
        <input
          class="btn-medusa manualSearchButton top-5 bottom-5"
          type="button"
          value="Refresh Results"
        />
         
        <input
          class="btn-medusa manualSearchButton top-5 bottom-5"
          type="button"
          value="Force Search"
        />
         
        <!---->
      </div>
       
      <vue-good-table-stub
        columnfilteroptions="[object Object]"
        columns="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
        groupoptions="[object Object]"
        mode="local"
        paginationoptions="[object Object]"
        responsive="true"
        rows="[object Object],[object Object],[object Object]"
        rowstyleclass="function () { [native code] }"
        searchoptions="[object Object]"
        selectoptions="[object Object]"
        sortoptions="[object Object]"
        styleclass="vgt-table condensed"
        theme=""
      />
    </div>
  </div>
</div>
`;
